File: tests/rosetta/transpiler/Java/100-prisoners.java
public class Main {

    static int[] shuffle(int[] xs) {
        int[] arr = xs;
        int i = 99;
        while (i > 0) {
            int j = _now() % (i + 1);
            int tmp = arr[i];
arr[i] = arr[j];
arr[j] = tmp;
            i = i - 1;
        }
        return arr;
    }

    static void doTrials(int trials, int np, String strategy) {
        int pardoned = 0;
        int t = 0;
        while (t < trials) {
            int[] drawers = new int[]{};
            int i = 0;
            while (i < 100) {
                drawers = java.util.stream.IntStream.concat(java.util.Arrays.stream(drawers), java.util.stream.IntStream.of(i)).toArray();
                i = i + 1;
            }
            drawers = shuffle(drawers);
            int p = 0;
            boolean success = true;
            while (p < np) {
                boolean found = false;
                if ((strategy.equals("optimal"))) {
                    int prev = p;
                    int d = 0;
                    while (d < 50) {
                        int this_ = drawers[prev];
                        if (this_ == p) {
                            found = true;
                            break;
                        }
                        prev = this_;
                        d = d + 1;
                    }
                } else {
                    boolean[] opened = new boolean[]{};
                    int k = 0;
                    while (k < 100) {
                        opened = appendBool(opened, false);
                        k = k + 1;
                    }
                    int d = 0;
                    while (d < 50) {
                        int n = _now() % 100;
                        while (opened[n]) {
                            n = _now() % 100;
                        }
opened[n] = true;
                        if (((Number)(drawers[n])).intValue() == p) {
                            found = true;
                            break;
                        }
                        d = d + 1;
                    }
                }
                if (!found) {
                    success = false;
                    break;
                }
                p = p + 1;
            }
            if (success) {
                pardoned = pardoned + 1;
            }
            t = t + 1;
        }
        double rf = (((Number)(pardoned)).doubleValue()) / (((Number)(trials)).doubleValue()) * 100.0;
        System.out.println("  strategy = " + strategy + "  pardoned = " + String.valueOf(pardoned) + " relative frequency = " + String.valueOf(rf) + "%");
    }

    static void main() {
        int trials = 1000;
        for (var np : new int[]{10, 100}) {
            System.out.println("Results from " + String.valueOf(trials) + " trials with " + String.valueOf(np) + " prisoners:\n");
            for (var strat : new String[]{"random", "optimal"}) {
                doTrials(trials, np, strat);
            }
        }
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        return rt.totalMemory() - rt.freeMemory();
    }

    static boolean[] appendBool(boolean[] arr, boolean v) {
        boolean[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }
}

File: tests/rosetta/transpiler/Kotlin/100-prisoners.kt
var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Int {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        _nowSeed.toInt()
    } else {
        System.nanoTime().toInt()
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun shuffle(xs: MutableList<Int>): MutableList<Int> {
    var arr: MutableList<Int> = xs
    var i: Int = 99
    while (i > 0) {
        val j: Int = _now() % (i + 1)
        val tmp: Int = arr[i]
        arr[i] = arr[j]
        arr[j] = tmp
        i = i - 1
    }
    return arr
}

fun doTrials(trials: Int, np: Int, strategy: String): Unit {
    var pardoned: Int = 0
    var t: Int = 0
    while (t < trials) {
        var drawers: MutableList<Int> = mutableListOf()
        var i: Int = 0
        while (i < 100) {
            drawers = run { val _tmp = drawers.toMutableList(); _tmp.add(i); _tmp } as MutableList<Int>
            i = i + 1
        }
        drawers = shuffle(drawers)
        var p: Int = 0
        var success: Boolean = true
        while (p < np) {
            var found: Boolean = false
            if (strategy == "optimal") {
                var prev: Int = p
                var d: Int = 0
                while (d < 50) {
                    val _this: Int = drawers[prev]
                    if (_this == p) {
                        found = true
                        break
                    }
                    prev = _this
                    d = d + 1
                }
            } else {
                var opened: MutableList<Boolean> = mutableListOf()
                var k: Int = 0
                while (k < 100) {
                    opened = run { val _tmp = opened.toMutableList(); _tmp.add(false); _tmp } as MutableList<Boolean>
                    k = k + 1
                }
                var d: Int = 0
                while (d < 50) {
                    var n: Int = _now() % 100
                    while ((opened[n]) as Boolean) {
                        n = _now() % 100
                    }
                    opened[n] = true
                    if (drawers[n] == p) {
                        found = true
                        break
                    }
                    d = d + 1
                }
            }
            if (!found) {
                success = false
                break
            }
            p = p + 1
        }
        if (success as Boolean) {
            pardoned = pardoned + 1
        }
        t = t + 1
    }
    val rf: Double = (pardoned.toDouble() / trials.toDouble()) * 100.0
    println(((((("  strategy = " + strategy) + "  pardoned = ") + pardoned.toString()) + " relative frequency = ") + rf.toString()) + "%")
}

fun user_main(): Unit {
    val trials: Int = 1000
    for (np in mutableListOf(10, 100)) {
        println(((("Results from " + trials.toString()) + " trials with ") + np.toString()) + " prisoners:\n")
        for (strat in mutableListOf("random", "optimal")) {
            doTrials(trials, np, strat)
        }
    }
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        user_main()
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
